### PR TITLE
BUILD-256: Conformance Tests for Shared Resource CSI Driver

### PR DESCRIPTION
**Pull request by prietyc123**
- Test feature file `csi-driver-e2e.feature` to test e2e functionality of csi driver
- step implementation details is in steps.py
- This PRs primary focus on sharedConfigMap, sharedSecret and metrics verification
- `make smoke` should run successfully on 4.10 openshift cluster

### CODE REVIEW

File: deploy/05-operator-subscription.yaml
Comment (gabemontero) [inline deploy/05-operator-subscription.yaml:1]: yeah @prietyc123 this will need to be moved to a test specific deploy directory vs. this runtime on

File: deploy/04-operatorgroup.yaml
Comment (gabemontero) [inline deploy/04-operatorgroup.yaml:1]: this is generic enough I'm not sure about it entirely, but if it is only needed for the postgresql subscription below, it too needs to be moved to a test specific deploy dir

**Review by gabemontero — CHANGES_REQUESTED** [review]
hey @prietyc123 I know this has been marked WIP, but I wanted to lay out a couple of starting points that I think need to change before you invest too much more

**Comment by adambkaplan** [comment]
> Now, of course, we log the lack of the config map. And I think we should add a new metric, if not with @alicerum's PR, with either a separate PR, or this PR on top of @alicerum 's PR, that track whether we have a valid config map, and fires an alert if we do not.

From an "operate first" perspective a metric + alert would make sense. I would do this work in a separate pull request.

We can also check if there are ways we can have the cluster storage operator mark itself degraded because a CSI driver has a bad configuration.

**Comment by prietyc123** [comment]
This pr states the functional test scenario for verifying shared resource csi driver behavior. This needs to go on top of pr https://github.com/openshift/csi-driver-shared-resource/pull/76 . Still most of the changes are req, once pr #76 merge in.

**Review by gabemontero — CHANGES_REQUESTED** [review]
@prietyc123 - I have asked @jitendar-singh in #76 to create a `test/bdd-tests` subfolder to house the base framework you've noted these changes will set on top of, vs. the `tests/functional` he currently has.

Track that discussion, and assuming that is where we land, I want any files you introduced with this PR to sit under `tests/bdd-tests`

File: test/csi-driver.feature
Comment (gabemontero) [inline test/csi-driver.feature:3]: 5 high level questions @prietyc123 @jitendar-singh 

1)  do you all envision descriptions like these being a replacement for what QE traditionally documented in polarion, or do you see these descriptions ultimately being items you drop into place into polarion? 

2) I am assuming we will see you all create separate `Feature` items for all the different scenarios that QA should cover for shared resources and 4.10.  Is that correct?

3) If so, do you all plan to track all of that under BUILD-256?

4) If so, can we expect separate PRs for each of the scenarios as you tackle them?

5) For items that have to be verified manually, like the metrics stuff through the console, should we expect those items to be described in polarion and presented to development for review?

@openshift/openshift-team-build-api FYI

Comment (prietyc123) [reply test/csi-driver.feature:3]: @gabemontero these description will replace polarion description that we do manually currently and then dev review that. Once we have these description ready, we have automated script in place to pull data to polarion. 
After the changes are done, team will review it in github itself. No need to go for polarion and stuff. 

> If so, can we expect separate PRs for each of the scenarios as you tackle them?

I will open separate pr for each and every e2e csi driver scenarios.

Comment (gabemontero) [reply test/csi-driver.feature:3]: OK @prietyc123 thanks for those clarifications.  So, next steps from my point of view:
- I have already asked @jitendar-singh to capture the answer to my questions in his #80 which he has told me will serve as the base for this PR.  I want our new / burgeoning process here documented in the repo's README, not just PR comment or slack comments ... he has updates in that PR which I'll look at shortly.  I'll cross reference as needed between what you noted here and his updates.
- separate PRs for each scenario is certainly reasonable; but we'll need to track somehow how far along you are in crafting items for each of the scenarios, so that we can properly tag items in Jira as "tc-approved".  How would you like to track ?  Minimally, a central list of the scenarios you will be going after seems prudent.

@adambkaplan FYI

**Review by jitendar-singh — COMMENTED** [review]
@prietyc123 please refactor the PR once my PR is merged, you need to add the feature files into smoke/features directory and for the yaml files create a new directory under smoke directory and have all the yaml in it.

**Comment by gabemontero** [comment]
/assign @jitendar-singh 

he should be the one looking at the python pieces, as he is most familiar with the BDD

I'll look at the feature files, yaml, overall scenarios, etc.

/approve

**Comment by gabemontero** [comment]
@prietyc123 is this still WIP, or should you remove that label at this point?

File: smoke/features/data/shareconfigmap.sh
Comment (gabemontero) [inline smoke/features/data/shareconfigmap.sh:33]: should you not also be creating a rolebinding that maps this role to the `default` serviceaccount?   I'm not seeing that being handled anywhere else

**Review by gabemontero — COMMENTED** [review]
As this is still marked WIP, I'm acknowledging that some of the items I'm about to note possible "are coming soon" :-) .

But for clarity, I'm going to list some items I am ultimately expecting with this PR
1) a feature file marked manual that details the step performed when you validated @alicerum 's metric work recently
2) we should see a shared secret included in this PR
3) cover configuration via https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#maintaining-the-configmap for validating the config options at https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#configuration-file ... hint, we minimally have a markdown link ref error in that section :-)
4) some amount of scale testing would be nice

Now, what exactly is meets min for this PR and completing QE review for BUILD-256,?  And what we can carry over to subsequent PR work that we manage after code freeeze, which most likely would not merge until after master branch switches to 4.11 work.

First blush, I think that 1) and 2) are meets min.  For 3), I think that is containable, but we can see how things progress during this sprint and if need be defer (we a fair amount of coverage for this in the PR e2e's.

For 4), I see that being a post feature freeze item.  So for this PR, minimally getting a feature file in place so that we have a placeholder for tracking, would be ideal for me.

And of course, we can discuss this, in addition to comments here in the PR, in scrum/office hours during this sprint.

thanks

File: smoke/features/data/shareconfigmap.sh
Comment (prietyc123) [reply smoke/features/data/shareconfigmap.sh:33]: Ohh yes, I missed it. I realised it after pushing the commit. So, I planned to push it in my next commit 🙂 . Thanks for pointing it as well.

File: smoke/features/csi-driver-e2e.feature
Comment (gabemontero) [inline smoke/features/csi-driver-e2e.feature:21]: so  you mention SharedSecrets up top, but there is no scenario that discussed creating a SharedSecret

Comment (gabemontero) [reply smoke/features/csi-driver-e2e.feature:21]: presumably that is coming "soon" @prietyc123 ?

Comment (prietyc123) [reply smoke/features/csi-driver-e2e.feature:21]: Yes @gabemontero  SharedSecret scenario will be added soon

**Comment by prietyc123** [comment]
> As this is still marked WIP, I'm acknowledging that some of the items I'm about to note possible "are coming soon" :-) .
> 
> But for clarity, I'm going to list some items I am ultimately expecting with this PR
> 
> 1. a feature file marked manual that details the step performed when you validated @alicerum 's metric work recently
> 2. we should see a shared secret included in this PR
> 3. cover configuration via https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#maintaining-the-configmap for validating the config options at https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#configuration-file ... hint, we minimally have a markdown link ref error in that section :-)
> 4. some amount of scale testing would be nice

@gabemontero I totally agree with you here. Although initially as part of this PR I am planning to cover the SharedSecrets, SharedConfigmaps and metrics verification part and later in chunks I will be adding stuffs like `hostpath volume mount`  scenarios worked by @akram and Rhel entitlement so on with separate PRs. However I would really like to know your thoughts on this. 

Secondly I also want to discuss about running these `.feature` file on PRs. How exactly we are thinking to run these files? Do we need to spun up another prow jobs or Existing one would work? We can discuss all these points in scrum/office hours as well.

**Comment by gabemontero** [comment]
> > As this is still marked WIP, I'm acknowledging that some of the items I'm about to note possible "are coming soon" :-) .
> > But for clarity, I'm going to list some items I am ultimately expecting with this PR
> > 
> > 1. a feature file marked manual that details the step performed when you validated @alicerum 's metric work recently
> > 2. we should see a shared secret included in this PR
> > 3. cover configuration via https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#maintaining-the-configmap for validating the config options at https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#configuration-file ... hint, we minimally have a markdown link ref error in that section :-)
> > 4. some amount of scale testing would be nice
> 
> @gabemontero I totally agree with you here. Although initially as part of this PR I am planning to cover the SharedSecrets, SharedConfigmaps and metrics verification part and later in chunks I will be adding stuffs like `hostpath volume mount` scenarios worked by @akram and Rhel entitlement so on with separate PRs. However I would really like to know your thoughts on this.

@prietyc123 I don't know what you mean by `hostpath volume mounts` scenarios,  but the stuff @akram was previously working on (BUILD-361 / BUILD-381) around "read only" is stuff I'm taking over as Jenkins Cpaas is taking too much time away from him.  I'm fine you handling that scenario separately from this one.

However, I will insist you take on the config map configuration change stuff either with this PR, or a follow up PR that is still part of BUILD-256.  Even if you do not automate, you should be able to try it manually in pretty short order.

As I indicated before, although we should have it eventually, I'm fine with scale testing being after BUILD-256 / 4.10 code freeze.

> 
> Secondly I also want to discuss about running these `.feature` file on PRs. How exactly we are thinking to run these files? Do we need to spun up another prow jobs or Existing one would work? We can discuss all these points in scrum/office hours as well.

As you know, we ended up talking about this in the retrospective today.  This is a complex undertaking that will require a separate R&D story to pursue, given our smoke/BDD dependencies around python etc. not being present in CI, and you cannot assume that you can just install stuff during CI like you do from the command line.  I'm sure as part of the notes @adambkaplan took during the retro meeting a story will be forthcoming for the 4.11 docket.

OK .... now moving on to my review of what you do have.

**Review by gabemontero — CHANGES_REQUESTED** [review]
got an initial set of stuff @prietyc123 

also, for the configuration scenario https://github.com/openshift/csi-driver-shared-resource/blob/master/docs/config.md#configuration-file the meets min manual test I want you to describe  is 
1) set refreshResources to false
2) confirm the shared resource driver pods in the `openshift-cluster-csi-drivers` project restart
3) use your basic sharedconfigmap test, but when you edit the configmap, the updates should *NOT* be reflected in the pod

File: hack/smoke-test.sh
Comment (gabemontero) [reply hack/smoke-test.sh:3]: I checked out you PR and ran `shellcheck` ... here are the items it reported for this one:

```
$ shellcheck smoke-test.sh 

In smoke-test.sh line 5:
  echo -e "\n!!!!!!!!!!!!!!!!!!!!\n${1}\n!!!!!!!!!!!!!!!!!!!!\n"
       ^-- SC3037: In POSIX sh, echo flags are undefined.


In smoke-test.sh line 49:
    if [[ $i == "testing-namespace"* ]]; then
       ^-- SC3010: In POSIX sh, [[ ]] is undefined.


In smoke-test.sh line 50:
        oc delete project $i
                          ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
        oc delete project "$i"

For more information:
  https://www.shellcheck.net/wiki/SC3010 -- In POSIX sh, [[ ]] is undefined.
  https://www.shellcheck.net/wiki/SC3037 -- In POSIX sh, echo flags are undef...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

File: smoke/features/data/pod.sh
Comment (gabemontero) [reply smoke/features/data/pod.sh:1]:
```
$ shellcheck pod.sh 

In pod.sh line 30:
if [ ${1} == "sharedconfigmap" ]; then
     ^--^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
if [ "${1}" == "sharedconfigmap" ]; then

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

File: smoke/features/data/rbac.sh
Comment (gabemontero) [reply smoke/features/data/rbac.sh:1]:
```
$ shellcheck rbac.sh 

In rbac.sh line 22:
if [ ${1} == "sharedconfigmap" ]; then
     ^--^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
if [ "${1}" == "sharedconfigmap" ]; then


In rbac.sh line 31:
oc create rolebinding shared-resource-my-shared --role=shared-resource-my-shared --serviceaccount=$(oc project -q):default
                                                                                                  ^--------------^ SC2046: Quote this to prevent word splitting.

For more information:
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

File: smoke/features/data/shareconfigmap.sh
Comment (gabemontero) [reply smoke/features/data/shareconfigmap.sh:1]:
```
$ shellcheck shareconfigmap.sh 

In shareconfigmap.sh line 17:
create_shared_configmap ${1}
                        ^--^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
create_shared_configmap "${1}"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

File: smoke/features/data/sharesecret.sh
Comment (gabemontero) [reply smoke/features/data/sharesecret.sh:1]: $ shellcheck sharesecret.sh 

In sharesecret.sh line 17:
create_shared_secret ${1}
                     ^--^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
create_shared_secret "${1}"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...

**Review by gabemontero — COMMENTED** [review]
The smoke/features/csi-driver-e2e.feature changes we previously talked about look good @prietyc123 

Next, to help facilitate what I was asking for with the various bash scripts, I posted the output from shellcheck for each of them.  Install shellcheck via `sudo dnf -y install shellcheck`, make the changes it asks for, make sure it runs clean against each new script, and push the updates here.

Then, I believe I heard in scrum that you were working on the feature file for the manual configuration change test we talked about.  So I'll sit tight and watch for you next PR update for that.

Lastly, while working on my BUILD-381 another important scenario that QE should cover occurred to me.  Namely, after a Pod is provision with a share resource volume, if the cluster admin removes permissions for the Pod service account to use the share, the contents from the volume should be removed.  Now, this happens on the controller relist, so it can take up to 10 minutes for the removal to occur.  And validation require that either your pod loops on displaying the contents of the volume, or you `oc rsh` into the Pod and confirm the contents are gone.  Bottom line, I think this is another manual test.  At least initially.  But if you think you can automate this, let me nkow.  But this is a key "user facing" feature that should fall under the coverage of QE.

There is a development e2e test in golang in this repo that automates this test if using that for reference helps.

File: hack/smoke-test.sh
Comment (prietyc123) [reply hack/smoke-test.sh:3]: Thanks @gabemontero for pointing it out. `shellcheck` error has been fixed now. If you could have a look 🙂

**Comment by prietyc123** [comment]
> The smoke/features/csi-driver-e2e.feature changes we previously talked about look good @prietyc123
> 
> Next, to help facilitate what I was asking for with the various bash scripts, I posted the output from shellcheck for each of them. Install shellcheck via `sudo dnf -y install shellcheck`, make the changes it asks for, make sure it runs clean against each new script, and push the updates here.

Done with the `shellcheck` error fixes

> 
> Then, I believe I heard in scrum that you were working on the feature file for the manual configuration change test we talked about. So I'll sit tight and watch for you next PR update for that.
> 
> Lastly, while working on my BUILD-381 another important scenario that QE should cover occurred to me. Namely, after a Pod is provision with a share resource volume, if the cluster admin removes permissions for the Pod service account to use the share, the contents from the volume should be removed. Now, this happens on the controller relist, so it can take up to 10 minutes for the removal to occur. And validation require that either your pod loops on displaying the contents of the volume, or you `oc rsh` into the Pod and confirm the contents are gone. Bottom line, I think this is another manual test. At least initially. But if you think you can automate this, let me nkow. But this is a key "user facing" feature that should fall under the coverage of QE.
> 
> There is a development e2e test in golang in this repo that automates this test if using that for reference helps.

Sure, I will try to run the mentioned scenario manually first and then will automated it. I think it can be doable. Not sure if that needs to be handled with this pr or another pr just after it merge would work.

File: smoke/features/csi-driver-e2e.feature
Comment (gabemontero) [inline smoke/features/csi-driver-e2e.feature:13]: I would mention in the scenario description here that you have included the refreshResources setting validation

**Review by gabemontero — CHANGES_REQUESTED** [review]
shellcheck updates look good @prietyc123 

just one minor change for your refreshResources addition

otherwise, I'll just wait for your update for the "data removed if permission removed"  scenario.  And yes, I think it is fine to just have a feature file with manual for this PR, and tackle automating it in a subsequent PR, separate from BUILD-256 (we'll mark BUILD-256 complete when this PR merges), and after 4.10 code freeze and when the main branch switches to 4.11 if need be.

**Comment by gabemontero** [comment]
/label px-approved
internal only QE item

**Comment by gabemontero** [comment]
/label docs-approved
internal only QE item

**Comment by gabemontero** [comment]
/label qe-approved
internal only QE item from QE

File: smoke/features/csi-driver-e2e.feature
Comment (gabemontero) [reply smoke/features/csi-driver-e2e.feature:13]: You missed making this update in your last commit @prietyc123

**Review by gabemontero — COMMENTED** [review]
I added some labels @prietyc123 

Otherwise, still looking for the rbac removed manual scenario documented, and that mention of refresh resources in the top line of the existing configmap scenario.

Also, now that @jitendar-singh is back, ideally I'd like for him to review the python pieces you have here and sign off on them in this PR.

File: smoke/features/csi-driver-e2e.feature
Comment (prietyc123) [reply smoke/features/csi-driver-e2e.feature]: Done

**Review by gabemontero — APPROVED** [review]
/lgtm

/hold

holding so that 

/assign @jitendar-singh 

I'd like a review from @jitendar-singh on the specific tests / python implementation.  I just focused on the scenario / test descriptions in the feature files.

**Comment by gabemontero** [comment]
aws failure

/retest

**Comment by gabemontero** [comment]
bump @jitendar-singh @prietyc123 

I only put the hold on here because I want @jitendar-singh to review the new python code from @prietyc123 
Once he does that and their are no pending updates, either of you can cancel the hold and this can merge.

**Pull request merged**
